### ISSUE TITLE
2.0 plan

### ISSUE THREAD

**Issue by sds100** [sds100/KeyMapper#181]
- drag and drop triggers and MULTIPLE actions into any order
- allow multiple of the same trigger in a sequence. e.g up, up, down, down
- provide option for old layout (Use modules?)
- double tap flag
- have chips. Show the icon for the action on the left and if an error, show error icon and glow red. Click to fix error. And if no error, click to test the action.
- have a separate develop-2.0 branch which 2.0 feature branches merge into.
- have chips span the whole screen with hamburger button on the side to drag and drop. Like Google maps when you rearrange stops
- automatically detect which device is sending the trigger and append it to the trigger chip with an •. E.g Vol Up • Logitech Keyboard. Test whether wired headphones are detected as a different source. key character map?
- click on trigger chips to change to go to "trigger options". Here, the user can change which device it listens from with option for all devices, whether it is a long press or double press
-  separate actions and triggers.
- have an easy to access help button in the menu
- move toolbar to bottom
- on home screen have expandable cards for each keymap. When expanded, show the trigger and action order. When collapsed, only describe the actions in order.
- put status layouts in an expandable card. automatically expand and collapse depending on whether something critical needs fixing
- DARK MODE! make it automatically toggle on and off on Q devices, have button to force it. Use DayNight theme with -night resources. Set the dark mode in the application object. https://material.io/components/android/theming/dark/
- use card style like Castro app
- have multiple action lists for each keymap. A list can have multiple constraints which mean it overrides other action lists. e.g a certain app must be open.
- have a checkbox to work if the screen is off. but must check only compatible keys are triggers. e.g volume up, down. and if there are other keys, tell the user why it cant work with the screen off. have button to automatically remove incompatible keys. 
- have preference to change the double press window. E.g the user has 1 second to triple press. Default should be 300ms
- when selecting  a keymap on the homescreen, have an option to clone it.
- show a message when the user firsts open ConfigKeyMapActivity telling them how to create a key map. Have an option to not show again. and show it until they click dont show again

Use InputDevice.getDevice(keyevent.deviceId).descriptor to get unique id for a device

Use val inputManager = ctx.getSystemService(Context.INPUT_SERVICE) as InputManager
        inputManager.registerInputDeviceListener() to listen for when devices are connected/disconnected in the accessibility service.

**Comment by sds100** [comment]
Sort these into individual issues then port into 2.0 project board.

**Comment by hoshsadiq** [comment]
I suggest that these features get coded up gradually rather than all in one go for 2.0 release. Sort of like a agile method. Otherwise it'll likely be overwhelming and bug fixing will take a lot longer.

**Comment by sds100** [comment]
Yeah. I see what you mean. A lot of the things in the list are only small tweaks once another one is finished so I don't think it is too much work. I'll sort through them and see which feature depends on another then decide which will go in the first release. Right now, I'm really busy with studying and probably can't (shouldn't) focus too much on the app.

**Comment by hoshsadiq** [comment]
No problem mate, good luck with your studies!

I'd be happy to help out but I know little about Android development, however, do let me know if there are any infrastructure and/or pipeline automation changes you'd like and I'd be happy to implement it.

**Comment by sds100** [comment]
That would be nice. I had a Jenkins server which would automatically upload
an alpha build to GitHub every Sunday. The changelog it made started
breaking and there wasn't much development going on so I stopped it.

**Comment by hoshsadiq** [comment]
No need for Jenkins, I'll set it up using either Travis CI or GitHub actions

**Comment by sds100** [comment]
I didn't know about GitHub actions. Is that new? I'm guessing it is like
Gitlab's pipeline feature?

**Comment by sudomain** [comment]
+1 for

> allow multiple of the same trigger in a sequence. e.g up, up, down, down

Will this allow for complex sequences such as: 

CTRL + (CTRL + F) which would mean press CTRL, release CTRL, then press CTRL and F at the same time

as well as CTRL + CTRL + F which would press CTRL, release CTRL, press CTRL again, release CTRL again, then press F

The key difference being the first sequence requires the (CTRL + F) keys to be pressed at the same time, while the second requires CTRL to be released before pressing F

**Comment by sds100** [comment]
I wasn't planning on doing that but that's a good idea. I need to find an intuitive way to design the UI without it being too complex and confusing. Any suggestions?

**Comment by sudomain** [comment]
I like the current "RECORD TRIGGER" button and the 5 seconds that the user has to input the key sequence. I can't imagine a sequence taking longer than 5 seconds to record and also being useful.

Correct me if I'm wrong, but it appears the app currently uses Android "chips" to display the trigger/shortcut. Maybe the chips can be replaced with a simple string that KeyMapper appends to when recording a trigger?

As for syntax, you could use `(KEY1 + KEY2)` to show that two or more keys need to be pressed at the same time. `KEY1 -> KEY2` could be used to show that KEY1 needs to be pressed and released before KEY2 is pressed in order to trigger an action.

**Comment by sds100** [comment]
Yeah. That's the design I've been working on already. The algorithm to detect these triggers will be very complicated and hard to make. I'll see how feasible it is when I start working on the app again.

**Comment by sds100** [comment]
Split up this issue into multiple tasks in the 2.0.0 project
